File: algorithms/cpp/merge_sort.cpp
#include <iostream>
using namespace std;
void Merge(int* A, int, int, int);
void Merge_Sort(int* A, int p, int r);
int main()
{
    int len, n, i;
    cout << "Array size: ";
    cin >> len;
    int a[len];
    for (i = 0; i < len; i++) {
        cout << "Enter element " << i + 1 << ": ";
        cin >> a[i];
    }
    cout << "Initial array: ";
    for (int i = 0; i < len; i++) {
        cout << a[i] << " ";
    }

    Merge_Sort(a, 0, len - 1);
    cout << "\nSorted array:  ";
    for (int i = 0; i < len; i++) {
        cout << a[i] << " ";
    }
    cout << "\n";
    return 0;
}
void Merge(int* arr, int p, int q, int r)
{
    int n1 = q - p + 1;
    int n2 = r - q;
    int L[n1 + 1];
    int R[n2 + 1];
    for (int i = 0; i < n1; i++)
        L[i] = arr[p + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[q + 1 + j];
    int i = 0;
    int j = 0;
    int n = 0;
    while (i != n1 && j != n2) {
        if (L[i] > R[j]) {
            arr[p + n] = R[j];
            j++;
        }
        else {
            arr[p + n] = L[i];
            i++;
        }
        n++;
    }
    while (j != n2) {
        arr[p + n] = R[j];
        j++;
        n++;
    }
    while (i != n1) {
        arr[p + n] = L[i];
        i++;
        n++;
    }
}
void Merge_Sort(int* arr, int p, int r)
{
    if (r > p) {
        int q;
        q = (p + r) / 2;
        Merge_Sort(arr, p, q);
        Merge_Sort(arr, q + 1, r);
        Merge(arr, p, q, r);
    }
}

File: algorithms/java/MergeSort.java
import java.util.Scanner;

public class MergeSort {

    public static void merge(int[] arr, int left, int mid, int right) {
        // Calculate the sizes of the two subarrays
        int size1 = mid - left + 1;
        int size2 = right - mid;

        // Create temporary arrays
        int[] leftArray = new int[size1];
        int[] rightArray = new int[size2];

        // Copy data to temporary arrays
        for (int i = 0; i < size1; ++i)
            leftArray[i] = arr[left + i];
        for (int j = 0; j < size2; ++j)
            rightArray[j] = arr[mid + 1 + j];

        // Merge the temporary arrays

        // Initial indexes of first and second subarrays
        int i = 0, j = 0;

        // Initial index of merged subarray array
        int k = left;
        while (i < size1 && j < size2) {
            if (leftArray[i] <= rightArray[j]) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of leftArray[], if any
        while (i < size1) {
            arr[k] = leftArray[i];
            i++;
            k++;
        }

        // Copy remaining elements of rightArray[], if any
        while (j < size2) {
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

   
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            // Find the middle point
            int mid = (left + right) / 2;

            // Recursive calls for left and right subarrays
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            // Merge the sorted subarrays
            merge(arr, left, mid, right);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Input prompt for array size
        System.out.print("Enter the size of the array: ");
        int size = scanner.nextInt();

        // Input prompt for array elements
        System.out.println("Enter the elements of the array:");
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = scanner.nextInt();
        }

        // Perform Merge Sort
        mergeSort(arr, 0, size - 1);

        // Output the sorted array
        System.out.println("Sorted array:");
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}


/*
Enter the size of the array: 5
Enter the elements of the array:
12 5 7 3 9
Sorted array:
3 5 7 9 12

 */